### PR TITLE
feat: Support basicAttConstraints X.509v3 extension

### PR DESCRIPTION
**Pull request by JonathanWilbur**
This PR adds support for the `basicAttConstraints` X.509v3 extension defined in ITU-T Recommendation X.509 for usage in attribute certificates. This is _different from, but similiar to_ the `basicConstraints` X.509v3 extension.

##### Checklist
- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by JonathanWilbur** [comment]
My PR is failing on style checks. I don't think it is possible for me to fix the styling problem: either I indent the arguments to those functions to column 61 and then the names overflow the 80-character column limit, or I indent them less to avoid it, which violates the rule. How do I proceed?

**Comment by JonathanWilbur** [comment]
Alternatively, I _could_ just put this in the same file as `basicConstraints`, since it uses the _exact same syntax_.

**Comment by mattcaswell** [comment]
> My PR is failing on style checks.

The style checking is new and somewhat experimental. Its currently coming up with quite a few false positives. Adding "style: waived"

File: test/certs/ext-basicAttConstraints.pem [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+MIIBjDCCAXigAwIBAgIDAQIDMAsGCSqGSIb3DQEBBTAAMCIYDzIwMjEwODMxMDIx
+MjQ5WhgPMjAyMTA4MzEwMjEyNDlaMAAwggEgMAsGCSqGSIb3DQEBAQOCAQ8AMIIB
+CgKCAQEAtnjLm1ts1hC4fNNt3UnQD9y73bDXgioTyWYSI3ca/KNfuTydjFTEYAmq
+nuGrBOUfgbmH3PRQ0AmpqljgWTb3d3K8H4UFvDWQTPSS21IMjm8oqd19nE5GxWir
+Gu0oDRzhWLHe1RZ7ZrohCPg/1Ocsy47QZuK2laFB0rEmrRWBmEYbDl3/wxf5XfqI
+qpOynJB02thXrTCcTM7Rz1FqCFt/ZVZB5hKY2S+CTdE9OIVKlr4WHMfuvUYeOj06
+GkwLFJHNv2tU+tovI3mYRxUuY4UupkS3MC+Otey7XKm1P+INjWWoegm6iCAt3Vus
+pVz+6pU2xgl3nrAVMQHB4fReQPH0pQIDAQABoxMwETAPBgNVHSkECDAGAQH/AgED
+MAsGCSqGSIb3DQEBBQMBAA==
+-----END CERTIFICATE-----
```
Comment (t8m) [inline test/certs/ext-basicAttConstraints.pem:11]: Please add the missing EOL

File: crypto/x509/build.info [hunk 2]
```diff
@@ -17,7 +17,7 @@ SOURCE[../../libcrypto]=\
         v3_asid.c v3_addr.c v3_tlsf.c v3_admis.c v3_no_rev_avail.c \
         v3_soa_id.c v3_no_ass.c v3_group_ac.c v3_single_use.c v3_ind_iss.c \
         x509_acert.c x509aset.c t_acert.c x_ietfatt.c v3_ac_tgt.c v3_sda.c \
-        v3_usernotice.c
+        v3_usernotice.c v3_bacons.c
```
Comment (t8m) [inline crypto/x509/build.info:20]: Nit: As bacon is a real word meaning something completely else, could you please avoid using this abbreviation? I.e. bacons -> battcons or something like that everywhere.

File: crypto/x509/v3_bacons.c [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+#include "internal/cryptlib.h"
+#include <openssl/asn1t.h>
+#include <openssl/conf.h>
+#include <openssl/x509v3.h>
+#include "x509_local.h"
+#include "ext_dat.h"
+
+static STACK_OF(CONF_VALUE) *i2v_OSSL_BASIC_ATTR_CONSTRAINTS(
+    X509V3_EXT_METHOD *method,
+    OSSL_BASIC_ATTR_CONSTRAINTS *bcons,
```
Comment (t8m) [inline crypto/x509/v3_bacons.c:19]: Here bcons is used, battcons would be better IMO

**Review by t8m — CHANGES_REQUESTED** [review]

**Comment by JonathanWilbur** [comment]
All feedback addressed and ready for review. I am pretty sure the one check failure (other than the code style, I mean) is unrelated to my changes.

**Comment by t8m** [comment]
All seems good except there is a merge commit. Could you please rebase to drop it?

**Comment by JonathanWilbur** [comment]
All feedback addressed and ready for review again. @nhorman would you be interested in reviewing this?

**Review by t8m — APPROVED** [review]

**Comment by JonathanWilbur** [comment]
@mattcaswell would you have any interest in reviewing this one?

**Comment by JonathanWilbur** [comment]
@t-j-h Would you be interested in reviewing this one, since it only needs one more approval?

**Comment by JonathanWilbur** [comment]
@t8m It seems like a lot of people that are usually responsive are not. This can't be a coincidence. Is everybody busy with the OpenSSL issues cleanup?

File: crypto/x509/v3_battcons.c [hunk 4]
```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 1999-2021 The OpenSSL Project Authors. All Rights Reserved.
```
Comment (FdaSilvaYY) [inline crypto/x509/v3_battcons.c:2]: nit: Just 2024

File: crypto/x509/v3_battcons.c [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+#include <openssl/conf.h>
+#include <openssl/x509v3.h>
+#include "x509_local.h"
+#include "ext_dat.h"
+
+static STACK_OF(CONF_VALUE) *i2v_OSSL_BASIC_ATTR_CONSTRAINTS(
+    X509V3_EXT_METHOD *method,
+    OSSL_BASIC_ATTR_CONSTRAINTS *battcons,
+    STACK_OF(CONF_VALUE)
+    *extlist);
```
Comment (FdaSilvaYY) [inline crypto/x509/v3_battcons.c:21]: nit : unneeded line break.

File: crypto/x509/v3_battcons.c [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    0, 0, 0, 0,
+    0, 0,
+    (X509V3_EXT_I2V) i2v_OSSL_BASIC_ATTR_CONSTRAINTS,
+    (X509V3_EXT_V2I)v2i_OSSL_BASIC_ATTR_CONSTRAINTS,
+    NULL, NULL,
+    NULL
+};
+
+ASN1_SEQUENCE(OSSL_BASIC_ATTR_CONSTRAINTS) = {
+        ASN1_OPT(OSSL_BASIC_ATTR_CONSTRAINTS, authority, ASN1_FBOOLEAN),
```
Comment (FdaSilvaYY) [inline crypto/x509/v3_battcons.c:40]: nit:  one indent is enough ?

File: crypto/x509/v3_battcons.c [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        ASN1_OPT(OSSL_BASIC_ATTR_CONSTRAINTS, pathlen, ASN1_INTEGER)
+} ASN1_SEQUENCE_END(OSSL_BASIC_ATTR_CONSTRAINTS)
+
+IMPLEMENT_ASN1_FUNCTIONS(OSSL_BASIC_ATTR_CONSTRAINTS)
+
+static STACK_OF(CONF_VALUE) *i2v_OSSL_BASIC_ATTR_CONSTRAINTS(
+    X509V3_EXT_METHOD *method,
+    OSSL_BASIC_ATTR_CONSTRAINTS *battcons,
+    STACK_OF(CONF_VALUE)
```
Comment (FdaSilvaYY) [inline crypto/x509/v3_battcons.c:49]: nit useless line break.

**Review by FdaSilvaYY — COMMENTED** [review]
👍

**Comment by t8m** [comment]
> @t8m It seems like a lot of people that are usually responsive are not. This can't be a coincidence. Is everybody busy with the OpenSSL issues cleanup?

More likely people are on vacations, etc.

**Review by nhorman — APPROVED** [review]

**Comment by JonathanWilbur** [comment]
@t8m @nhorman @FdaSilvaYY , all feedback addressed and this is ready for review again. (Just a few trivial style changes.)

**Comment by JonathanWilbur** [comment]
Is this able to receive the `approval: done` label yet?

**Comment by nhorman** [comment]
yep, has acks from myself and @t8m.  Not sure why the automation didn't flag it.

**Comment by nhorman** [comment]
pushed to master, thank you!

**Pull request closed without merging**